# Relatório de Análise 5

## Tratamento de Dados Faltantes

In [71]:
import pandas as pd

In [72]:
dados = pd.read_csv('dados/aluguel_residencial.csv', sep = ';')

In [73]:
dados

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
4      Apartamento              Cachambi        2      0       0    50   
...            ...                   ...      ...    ...     ...   ...   
21821  Apartamento                 Méier        2      0       0    70   
21822    Quitinete                Centro        0      0       0    27   
21823  Apartamento           Jacarepaguá        3      1       2    78   
21824  Apartamento  São Francisco Xavier        2      1       0    48   
21825  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         0.0   0.0  
2       800.0       390.0  20.0  
3       800.0       230.0   0.0  
4      1300.0       301.0  17.0  
...       ...         ...   ...  
21821   900.0       490.0  48.0  
21822   800.0       350.0  25.0  
21823  1800.0       800.0  40.0  
21824  1400.0       509.0  37.0  
21825  3000.0       760.0   0.0  

[21826 rows x 9 columns]

In [74]:
dados.head(10)

Tipo               Bairro  Quartos  Vagas  Suites  Area  \
0           Quitinete           Copacabana        1      0       0    40   
1                Casa      Jardim Botânico        2      0       1   100   
2         Apartamento               Centro        1      0       0    15   
3         Apartamento         Higienópolis        1      0       0    48   
4         Apartamento             Cachambi        2      0       0    50   
5  Casa de Condomínio      Barra da Tijuca        5      4       5   750   
6  Casa de Condomínio                Ramos        2      2       0    65   
7         Apartamento               Grajaú        2      1       0    70   
8         Apartamento  Lins de Vasconcelos        3      1       1    90   
9         Apartamento           Copacabana        1      0       1    40   

     Valor  Condominio  IPTU  
0   1700.0       500.0  60.0  
1   7000.0         0.0   0.0  
2    800.0       390.0  20.0  
3    800.0       230.0   0.0  
4   1300.0       301.0  17.0  
5  22000.0         0.0   0.0  
6   1000.0         0.0   0.0  
7   1500.0       642.0  74.0  
8   1500.0       455.0  14.0  
9   2000.0       561.0  50.0

## Checando os dados faltantes

Temos 2 formas de fazer a checagem.
Usando isnull() ou o notnull()

In [75]:
# Usando isnull()
dados.isnull()

Tipo  Bairro  Quartos  Vagas  Suites   Area  Valor  Condominio   IPTU
0      False   False    False  False   False  False  False       False  False
1      False   False    False  False   False  False  False       False  False
2      False   False    False  False   False  False  False       False  False
3      False   False    False  False   False  False  False       False  False
4      False   False    False  False   False  False  False       False  False
...      ...     ...      ...    ...     ...    ...    ...         ...    ...
21821  False   False    False  False   False  False  False       False  False
21822  False   False    False  False   False  False  False       False  False
21823  False   False    False  False   False  False  False       False  False
21824  False   False    False  False   False  False  False       False  False
21825  False   False    False  False   False  False  False       False  False

[21826 rows x 9 columns]

In [76]:
# Usando o notnull()
dados.notnull()

Tipo  Bairro  Quartos  Vagas  Suites  Area  Valor  Condominio  IPTU
0      True    True     True   True    True  True   True        True  True
1      True    True     True   True    True  True   True        True  True
2      True    True     True   True    True  True   True        True  True
3      True    True     True   True    True  True   True        True  True
4      True    True     True   True    True  True   True        True  True
...     ...     ...      ...    ...     ...   ...    ...         ...   ...
21821  True    True     True   True    True  True   True        True  True
21822  True    True     True   True    True  True   True        True  True
21823  True    True     True   True    True  True   True        True  True
21824  True    True     True   True    True  True   True        True  True
21825  True    True     True   True    True  True   True        True  True

[21826 rows x 9 columns]

In [77]:
# Verificando as informações do DataFrame
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21826 entries, 0 to 21825
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        21826 non-null  object 
 1   Bairro      21826 non-null  object 
 2   Quartos     21826 non-null  int64  
 3   Vagas       21826 non-null  int64  
 4   Suites      21826 non-null  int64  
 5   Area        21826 non-null  int64  
 6   Valor       21826 non-null  float64
 7   Condominio  21826 non-null  float64
 8   IPTU        21826 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.5+ MB


É possível identificar que nas variáveis: Valor, Condomínio e IPTU há valores nulos

In [78]:
# A coluna 'Valor' é a mais importante para nós, então vamos checar seus valores nulos
dados['Valor'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
21821    False
21822    False
21823    False
21824    False
21825    False
Name: Valor, Length: 21826, dtype: bool

In [79]:
# Passando para o nosso DataFrame
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

In [80]:
# Tamanho do DF antes de eliminarmos os nulos
A = dados.shape[0]
A

21826

In [81]:
"""
Eliminando os valores nulos.
Afinal só vamos trabalhar com imóveis que nos informem os valores dos aluguéis.
"""

# Usando o dropna()
A = dados.shape[0]
dados.dropna(subset = ['Valor'])
B = dados.shape[0]
A - B

0

In [82]:
# Precisamos passar o inplace para executar a função diretamente
A = dados.shape[0]
dados.dropna(subset = ['Valor'], inplace = True)
B = dados.shape[0]
A - B

0

In [83]:
# Se procurarmos os nulos agora, não encontraremos mais
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

## Tratamento Condicional

Temos as variáveis Condomínio e IPTU com alguns problemas. No caso de Condomínio, os dados nulos nem sempre são incorretos, afinal no caso de uma casa, realmente não há esse tipo de tarifa. Para este caso, devemos considerar a variável Tipo: caso o imóvel seja um apartamento, os dados nulos serão excluídos.    
Porque não interessa para nós um apartamento com valor de condomínio nulo.


In [84]:
# Primeiramente, verificaremos quantas assinaturas nulas temos em Condominio:
dados[dados['Condominio'].isnull()].shape[0]

0

In [85]:
# Agora precisamos filtrar os dados de acordo com o tipo Apartamento e que tenham condomínio nulo.
selecao = (dados['Tipo'] == 'Apartamento') & (dados['Condominio'].isnull())
selecao

0        False
1        False
2        False
3        False
4        False
         ...  
21821    False
21822    False
21823    False
21824    False
21825    False
Length: 21826, dtype: bool

__Nós selecionamos os apartamentos que tem condomínio nulo, mas o quereremos mesmo é removê-los__

Contudo, se nos atentarmos a nossa seleção, verificaremos que ela nos fornece exatamente as assinaturas que deverão ser removidas do banco de dados. Reaproveitaremos a estrutura já utilizada anteriormente neste notebook, e passaremos a seleção que criamos (selecao). Para eliminar as assinaturas nulas selecionadas ao invés de coletá-las, iremos adicionar ~ em selecao, que inverte a Series booleana. 

In [86]:
# Removendo os apartamentos com condomínio nulo
A = dados.shape[0]
dados = dados[~selecao] # com o til npos conseguimos inverter a seleção
B = dados.shape[0]
A - B

0

In [87]:
dados[dados['Condominio'].isnull()].shape[0]

0

__Vamos manter no banco os itens com condomínio e IPTU nulos, mas vamos atribuir um valor a eles.
Vamos atribuir o valor 0.__

In [88]:
# Usando fillna()
dados.fillna((0), inplace = True)
dados

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
4      Apartamento              Cachambi        2      0       0    50   
...            ...                   ...      ...    ...     ...   ...   
21821  Apartamento                 Méier        2      0       0    70   
21822    Quitinete                Centro        0      0       0    27   
21823  Apartamento           Jacarepaguá        3      1       2    78   
21824  Apartamento  São Francisco Xavier        2      1       0    48   
21825  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         0.0   0.0  
2       800.0       390.0  20.0  
3       800.0       230.0   0.0  
4      1300.0       301.0  17.0  
...       ...         ...   ...  
21821   900.0       490.0  48.0  
21822   800.0       350.0  25.0  
21823  1800.0       800.0  40.0  
21824  1400.0       509.0  37.0  
21825  3000.0       760.0   0.0  

[21826 rows x 9 columns]

Outra maneira de resolver a questão, é criar um dicionário que indicará os valores das variáveis. Podemos, inclusive, inserir valores diferentes para ambas as variáveis.

In [89]:
# Outra forma de fazer: usando fillna() e dicionário
dados = dados.fillna({'Condominio': 0, 'IPTU': 0})

In [90]:
dados

Tipo                Bairro  Quartos  Vagas  Suites  Area  \
0        Quitinete            Copacabana        1      0       0    40   
1             Casa       Jardim Botânico        2      0       1   100   
2      Apartamento                Centro        1      0       0    15   
3      Apartamento          Higienópolis        1      0       0    48   
4      Apartamento              Cachambi        2      0       0    50   
...            ...                   ...      ...    ...     ...   ...   
21821  Apartamento                 Méier        2      0       0    70   
21822    Quitinete                Centro        0      0       0    27   
21823  Apartamento           Jacarepaguá        3      1       2    78   
21824  Apartamento  São Francisco Xavier        2      1       0    48   
21825  Apartamento                Leblon        2      0       0    70   

        Valor  Condominio  IPTU  
0      1700.0       500.0  60.0  
1      7000.0         0.0   0.0  
2       800.0       390.0  20.0  
3       800.0       230.0   0.0  
4      1300.0       301.0  17.0  
...       ...         ...   ...  
21821   900.0       490.0  48.0  
21822   800.0       350.0  25.0  
21823  1800.0       800.0  40.0  
21824  1400.0       509.0  37.0  
21825  3000.0       760.0   0.0  

[21826 rows x 9 columns]

In [91]:
# Checando se ainda há dados nulos na variável Condominio
dados[dados['Condominio'].isnull()].shape[0]

0

In [92]:
# Checando se ainda há dados nulos na variável IPTU
dados[dados['IPTU'].isnull()].shape[0]

0

In [93]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21826 entries, 0 to 21825
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        21826 non-null  object 
 1   Bairro      21826 non-null  object 
 2   Quartos     21826 non-null  int64  
 3   Vagas       21826 non-null  int64  
 4   Suites      21826 non-null  int64  
 5   Area        21826 non-null  int64  
 6   Valor       21826 non-null  float64
 7   Condominio  21826 non-null  float64
 8   IPTU        21826 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.7+ MB


Temos, portanto, 21826 registros e nenhum deles é nulo. Para finalizarmos essa etapa, devemos salvar este arquivo, de forma que o DataFrame seja atualizado com as modificações que realizamos. Para tanto, escreveremos - lembrando que não queremos incluir o índice:

In [94]:
dados.to_csv('dados/aluguel_residencial.csv', sep = ';', index = False)

## Exercícios

#### 01 - Revisando o processo

Utilizando os recursos aprendidos nesta aula, julgue o conjunto de operações de limpeza do DataFrame abaixo:

In [95]:
imoveis = pd.DataFrame([['Apartamento', None, 970, 68],
                        ['Apartamento', 2000, 878, 112],
                        ['Casa', 5000, None, 500],
                        ['Apartamento', None, 1010, 170],
                        ['Apartamento', 1500, 850, None],
                        ['Casa', None, None, None],
                        ['Apartamento', 2000, 878, None],
                        ['Apartamento', 1550, None, 228],
                        ['Apartamento', 2500, 880, 195]],
                        columns = ['Tipo', 'Valor', 'Condominio', 'IPTU'])
imoveis

Tipo   Valor  Condominio   IPTU
0  Apartamento     NaN       970.0   68.0
1  Apartamento  2000.0       878.0  112.0
2         Casa  5000.0         NaN  500.0
3  Apartamento     NaN      1010.0  170.0
4  Apartamento  1500.0       850.0    NaN
5         Casa     NaN         NaN    NaN
6  Apartamento  2000.0       878.0    NaN
7  Apartamento  1550.0         NaN  228.0
8  Apartamento  2500.0       880.0  195.0

1) Elimina os registros que não apresentam a variável Valor:
```
Tipo	Valor	Condominio	IPTU
1	Apartamento	2000.0	878.0	112.0
2	Casa	5000.0	NaN	500.0
4	Apartamento	1500.0	850.0	NaN
6	Apartamento	2000.0	878.0	NaN
7	Apartamento	1550.0	NaN	228.0
8	Apartamento	2500.0	880.0	195.0
```

In [96]:
imoveis.dropna(subset = ['Valor'], inplace = True)
imoveis

# Verdadeiro

Tipo   Valor  Condominio   IPTU
1  Apartamento  2000.0       878.0  112.0
2         Casa  5000.0         NaN  500.0
4  Apartamento  1500.0       850.0    NaN
6  Apartamento  2000.0       878.0    NaN
7  Apartamento  1550.0         NaN  228.0
8  Apartamento  2500.0       880.0  195.0

2) Elimina os imóveis do tipo Apartamento que não apresentam valor Condominio:
```
	Tipo	Valor	Condominio	IPTU
1	Apartamento	2000.0	878.0	112.0
2	Casa	5000.0	NaN	500.0
4	Apartamento	1500.0	850.0	NaN
6	Apartamento	2000.0	878.0	NaN
8	Apartamento	2500.0	880.0	195.0
```

In [97]:
imoveis = pd.DataFrame([['Apartamento', None, 970, 68],
                        ['Apartamento', 2000, 878, 112],
                        ['Casa', 5000, None, 500],
                        ['Apartamento', None, 1010, 170],
                        ['Apartamento', 1500, 850, None],
                        ['Casa', None, None, None],
                        ['Apartamento', 2000, 878, None],
                        ['Apartamento', 1550, None, 228],
                        ['Apartamento', 2500, 880, 195]],
                        columns = ['Tipo', 'Valor', 'Condominio', 'IPTU'])
imoveis

Tipo   Valor  Condominio   IPTU
0  Apartamento     NaN       970.0   68.0
1  Apartamento  2000.0       878.0  112.0
2         Casa  5000.0         NaN  500.0
3  Apartamento     NaN      1010.0  170.0
4  Apartamento  1500.0       850.0    NaN
5         Casa     NaN         NaN    NaN
6  Apartamento  2000.0       878.0    NaN
7  Apartamento  1550.0         NaN  228.0
8  Apartamento  2500.0       880.0  195.0

In [98]:
selecao = (imoveis['Tipo'] == 'Apartamento') & (imoveis['Condominio'].isnull())
imoveis = imoveis[selecao]
imoveis

# Falso. Ele exibe onde tem faltante. Mas queremos onde não tem faltante.

Tipo   Valor  Condominio   IPTU
7  Apartamento  1550.0         NaN  228.0

__A alternativa 2 está errada. O que queremos neste caso é eliminar os apartamentos que apresentam condomínio nulo. Para isso, precisaríamos modificar a segunda linha do código acima:__
```
imoveis = imoveis[~selecao]
```

3) Substitui os valores faltantes que restam nas variáveis Condominio e IPTU por zero:
```
	Tipo	Valor	Condominio	IPTU
1	Apartamento	2000.0	878.0	112.0
2	Casa	5000.0	0.0	500.0
4	Apartamento	1500.0	850.0	0.0
6	Apartamento	2000.0	878.0	0.0
8	Apartamento	2500.0	880.0	195.0
```

In [99]:
imoveis = pd.DataFrame([['Apartamento', None, 970, 68],
                        ['Apartamento', 2000, 878, 112],
                        ['Casa', 5000, None, 500],
                        ['Apartamento', None, 1010, 170],
                        ['Apartamento', 1500, 850, None],
                        ['Casa', None, None, None],
                        ['Apartamento', 2000, 878, None],
                        ['Apartamento', 1550, None, 228],
                        ['Apartamento', 2500, 880, 195]],
                        columns = ['Tipo', 'Valor', 'Condominio', 'IPTU'])
imoveis

Tipo   Valor  Condominio   IPTU
0  Apartamento     NaN       970.0   68.0
1  Apartamento  2000.0       878.0  112.0
2         Casa  5000.0         NaN  500.0
3  Apartamento     NaN      1010.0  170.0
4  Apartamento  1500.0       850.0    NaN
5         Casa     NaN         NaN    NaN
6  Apartamento  2000.0       878.0    NaN
7  Apartamento  1550.0         NaN  228.0
8  Apartamento  2500.0       880.0  195.0

In [100]:
imoveis = imoveis.fillna({'Condominio': 0, 'IPTU': 0})
imoveis

# Verdadeiro

Tipo   Valor  Condominio   IPTU
0  Apartamento     NaN       970.0   68.0
1  Apartamento  2000.0       878.0  112.0
2         Casa  5000.0         0.0  500.0
3  Apartamento     NaN      1010.0  170.0
4  Apartamento  1500.0       850.0    0.0
5         Casa     NaN         0.0    0.0
6  Apartamento  2000.0       878.0    0.0
7  Apartamento  1550.0         0.0  228.0
8  Apartamento  2500.0       880.0  195.0

4) Reconstrói o índice do DataFrame resultante:
```
	Tipo	Valor	Condominio	IPTU
0	Apartamento	2000.0	878.0	112.0
1	Casa	5000.0	0.0	500.0
2	Apartamento	1500.0	850.0	0.0
3	Apartamento	2000.0	878.0	0.0
4	Apartamento	2500.0	880.0	195.0
```

In [101]:
imoveis.index = range(imoveis.shape[0])
imoveis

# Verdadeiro

Tipo   Valor  Condominio   IPTU
0  Apartamento     NaN       970.0   68.0
1  Apartamento  2000.0       878.0  112.0
2         Casa  5000.0         0.0  500.0
3  Apartamento     NaN      1010.0  170.0
4  Apartamento  1500.0       850.0    0.0
5         Casa     NaN         0.0    0.0
6  Apartamento  2000.0       878.0    0.0
7  Apartamento  1550.0         0.0  228.0
8  Apartamento  2500.0       880.0  195.0

#### 02 - Preenchendo dados faltantes

Suponha o seguinte conjunto de dados:

In [102]:
atletas = pd.DataFrame([['Marcos', 9.62], ['Pedro', None], ['João', 9.69],
                        ['Beto', 9.72], ['Sandro', None], ['Denis', 9.69],
                        ['Ary', None], ['Carlos', 9.74]],
                        columns = ['Corredor', 'Melhor Tempo'])
atletas

Corredor  Melhor Tempo
0   Marcos          9.62
1    Pedro           NaN
2     João          9.69
3     Beto          9.72
4   Sandro           NaN
5    Denis          9.69
6      Ary           NaN
7   Carlos          9.74

In [125]:
# Esta foi a resposta do gabarito do exercício
atletas.fillna(atletas.mean())

/tmp/ipykernel_704/178027887.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  atletas.fillna(atletas.mean())


Corredor  Melhor Tempo
0   Marcos         9.620
1    Pedro         9.692
2     João         9.690
3     Beto         9.720
4   Sandro         9.692
5    Denis         9.690
6      Ary         9.692
7   Carlos         9.740

#### Mas retornou o FutureWarning abaixo:
```
/tmp/ipykernel_704/178027887.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  atletas.fillna(atletas.mean())
```

__Consegui resolver na célula seguinte e precisei usar o transpose(), que também pode ser usado como T:__

In [150]:
# Sem o transpose o DataFrame foi gerado na horizontal
df = pd.DataFrame([atletas['Corredor'], atletas['Melhor Tempo'].fillna(atletas['Melhor Tempo'].mean())])
df

0      1     2     3       4      5      6       7
Corredor      Marcos  Pedro  João  Beto  Sandro  Denis    Ary  Carlos
Melhor Tempo    9.62  9.692  9.69  9.72   9.692   9.69  9.692    9.74

In [149]:
# Com o transpose a posição foi corrigida
df = pd.DataFrame([atletas['Corredor'], atletas['Melhor Tempo'].fillna(atletas['Melhor Tempo'].mean())]).transpose()
df
# ou
#df.T

Corredor Melhor Tempo
0   Marcos         9.62
1    Pedro        9.692
2     João         9.69
3     Beto         9.72
4   Sandro        9.692
5    Denis         9.69
6      Ary        9.692
7   Carlos         9.74

## Fim